tools/swarm/src/lib.rs

```diff
@@ -0,0 +1,418 @@
+
+impl<'a, 'temp> ImageSettings<'a> {
+    fn new(
+        name: &'a str,
+        build_dir: Option<&std::path::Path>,
+        ignore_cache: bool,
+        target_dir: &'temp path::AbsolutePath,
+    ) -> Result<Self, Error> {
+        Ok(Self {
+            name,
+            build_dir: build_dir
+                .map(path::AbsolutePath::new)
+                .transpose()?
+                .map(|dir| dir.relative_to(target_dir))
+                .transpose()?,
+            ignore_cache,
+        })
+    }
+}
+
+impl<'a> Swarm<'a> {
+    /// Creates a new Swarm generator.
+    #[allow(clippy::too_many_arguments, clippy::missing_errors_doc)]
+    pub fn new(
+        count: std::num::NonZeroU16,
+        seed: Option<&'a [u8]>,
+        healthcheck: bool,
+        config_dir: &'a std::path::Path,
+        image: &'a str,
+        build_dir: Option<&'a std::path::Path>,
+        ignore_cache: bool,
+        target_path: &'a std::path::Path,
+    ) -> Result<Self, Error> {
+        if target_path.is_dir() {
+            return Err(Error::TargetFileIsADirectory);
+        }
+        let target_path = path::AbsolutePath::new(target_path)?;
+        let target_dir = target_path.parent().ok_or(Error::NoTargetDirectory)?;
+        Ok(Self {
+            peer: PeerSettings::new(count, seed, healthcheck, config_dir, &target_dir)?,
+            image: ImageSettings::new(image, build_dir, ignore_cache, &target_dir)?,
+            target_path,
+        })
+    }
+
+    /// Builds the schema.
+    #[allow(clippy::missing_errors_doc)]
+    pub fn build(&self) -> schema::DockerCompose {
+        schema::DockerCompose::new(&self.image, &self.peer)
+    }
+
+    /// Returns the absolute target file path.
+    pub fn absolute_target_path(&self) -> &std::path::Path {
+        self.target_path.as_ref()
+    }
+}
+
+impl From<path::Error> for Error {
+    fn from(error: path::Error) -> Self {
+        Self::PathConversion(error)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #![allow(clippy::too_many_lines, clippy::needless_raw_string_hashes)]
+
+    use crate::Swarm;
+
+    const IMAGE: &str = "hyperledger/iroha:dev";
+    const PEER_CONFIG_PATH: &str = "./configs/swarm";
+    const TARGET_PATH: &str = "./configs/swarm/docker-compose.yml";
+
+    fn build_as_string(
+        count: std::num::NonZeroU16,
+        healthcheck: bool,
+        build_dir: Option<&str>,
+        ignore_cache: bool,
+        banner: Option<&[&str]>,
+    ) -> String {
+        let mut buffer = Vec::new();
+        Swarm::new(
+            count,
+            Some(&[]),
+            healthcheck,
+            PEER_CONFIG_PATH.as_ref(),
+            IMAGE,
+            build_dir.map(str::as_ref),
+            ignore_cache,
+            TARGET_PATH.as_ref(),
+        )
+        .unwrap()
+        .build()
+        .write(&mut buffer, banner)
+        .unwrap();
+        String::from_utf8(buffer).unwrap()
+    }
+
+    #[test]
+    fn single_build_banner() {
+        expect_test::expect!([r##"
```

Comment:
`expect_file!` didn't work? I think it could be better to move these snippets to dedicated files

Reply:
Yeah, the yaml generator doesn't indent lists. They get properly indented on reformatting if placed in a file, breaking the tests. https://github.com/hyperledger/iroha/pull/4719#discussion_r1637611496